TOOLS.rst

```diff
@@ -0,0 +1,201 @@
+=======================================
+
+The reason there are two different tool runners is that they have very
+different purposes:
+
+**invoke** is used as general purpose, platform independent way of
+defining and running "tasks relevant for development" (reminiscent of
+the *makefiles* of olden times, commonly used for software development
+on Unix systems). These tasks can be something like running a linter,
+type checker, test runner or code formatter, building documentation,
+cleaning the repository of temporary files, building a package or
+updating the list of contributors from the git log.
+
+**tox** is tuned towards a specific purpose, which is to run
+specifically QA tools (something like linter, type checker, test
+runner or code formatter) in a clean environment, with a fresh install
+of code, across as many supported python versions as possible. ``tox``
+can also be used in CI tools, such a Github Actions, to specify what
+to run as part of the CI, which eliminates the need to an extra
+configuration.
+
+Both tools share the purpose of hiding the details of the way the
+tools should be run (arguments, settings etc.) and defining a single
+place to define those.
+
+From the definitions above, however, it is obvious that the QA tools
+may be run by both. This is ok, because the purposes are different.
+But it raises the important concern of *where* the one true place to
+define the arguments and settings are for those. See definitions of
+that in the following section.
+
+Where are settings and arguments kept
+-------------------------------------
+
+For settings and arguments for tools the following rules apply:
+
+* **setup.cfg** If the tool support defining them in the shared
+  configuration file ``setup.cfg``, then they belong there no-matter
+  how they are run. If not (settings may only be available at the
+  command line), see the next points.
+* **tox.ini** If the tool is run by both ``tox`` and ``invoke``, then
+  the settings belong in the ``tox`` configuration file and ``invoke``
+  will have to read them from there
+* **tasks.py** If the tool is ``invoke`` specific, then obviously the
+  configuration goes into its "configuration file" ``tasks.py``
+
+Non-tool related "settings"
+```````````````````````````
+
+**Package metadata** goes into src/ixdat/__init__.py and README.rst
+and all remaining information about how to create a package goes into
+setup.py
+
+**requirements** goes into the two requirements files;
+``requirements.txt`` for package requirements and
+``requirements-dev.txt`` for development requirements.
+
+Git hooks
+=========
+
+The version control system ``git`` has the option of having check done
+before certain important actions, like a commit or a push. These are
+referred to as ``hooks``. For xidat a pre-push is recommended (as
+opposed to a pre-commit hook, which take time to run the tests on
+every commit). The pre-push hooks are located in ``tools/hooks``.
```

Comment:
Yeah. There isn't actually a git hook for that, since building the package isn't related to git. But I see your point. Normally this is done by settings up a CI tool (it is the next thing that I will propose) so that the QA tools are run automatically on a server, on the HEAD of master and on the head of every single PR ever, so that you will know immediately whether the code passes QA. In that case, actually building a package will then happen on code you likely checked our and package without modification, and that you therefore know passes QA.

You could build a soft hook by making the build of the package a task in invoke, that will run tox first on the available environments and exist with error if they do not pass, before actually building the package. That is commonly regarded a sufficient check. What do you think?

Reply:
I like that idea! `invoke build`?

Reply:
Yeah :)

I mean, the whole process consist of 3 steps maybe, with sub-steps. It could be something like (just brainstorming here):

Release:
 * Check that there are release notes in the documentation, if we use such
 * Update the version number and commit
 * Tag the release in git (using "v....." tags in git is common practice)
 * Push both the new commit and the tag

Build:
 * Perform local QA
 * Build
 
Upload:
 * Upload to PyPI

In principle any one of these could be automated and you could break this up in 3, 2 or 1 (depending on temper) tasks in invoke and have the whole release process scripted. Of course there are things which cannot be scripted like actually writing the release notes and making a new milestone on github of we use those, so we should still have a release procedure to follow, but it can be made significantly shorter and more re-producable.

Let's look into that separately.

PS: If you are interested you can read the SoCo release procedure here. It is missing some thing compared to above, because we used to have our CI tool (TravisCI) build and publish the packages, but that service is no longer available, at least not from there.